lib/ansible/modules/system/osx_defaults.py

```diff
@@ -15,413 +16,975 @@
+      (because they contains colons and that mess with the syntax).
+    - All dates need to be quoted.
+    - First level (not nested) quoted boolean, integers and floats are
+      converted to boolean, integer and float unless type is specified.
+    - List and dict can't be given as value in key=argument form.
+    - Binary data need to be encoded in base64.
 '''
 
 EXAMPLES = '''
+# Show debug menu in Safari. Specifying the type.
 - osx_defaults:
     domain: com.apple.Safari
     key: IncludeInternalDebugMenu
     type: bool
     value: true
     state: present
-
+# Set Measurement unit to cm for all apps (NSGlobalDomain).
 - osx_defaults:
     domain: NSGlobalDomain
     key: AppleMeasurementUnits
     type: string
     value: Centimeters
     state: present
-
+# Set clock visibility on screensaver for all users on current host.
 - osx_defaults:
     domain: com.apple.screensaver
     host: currentHost
+    any_user: true
     key: showClock
     type: int
     value: 1
-
+# Set Measurement unit to cm for all apps (NSGlobalDomain). No type specified.
 - osx_defaults:
     key: AppleMeasurementUnits
     type: string
     value: Centimeters
-
+# Set AppleLanguages to a list of languages.
 - osx_defaults:
     key: AppleLanguages
-    type: array
-    value:
-      - en
-      - nl
-
+    value: ["en", "nl"]
+# Remove a key.
 - osx_defaults:
     domain: com.geekchimp.macable
     key: ExampleKeyToRemove
     state: absent
+# Set a date. Dates need to be quoted. Timezone can be omitted. Local timezone will be use in this case.
+- osx_defaults:
+    domain: com.geekchimp.macable
+    key: SomeDate
+    value: "2002-12-15 02:59:43Z"
+# Set the nested key arrangeBy in abbreviated form (nested key need to be quoted).
+- osx_defaults: { domain: com.apple.finder, key: 'DesktopViewSettings:IconViewSettings:arrangeBy', value: dateModified }
+# Set ListViewSettings key to complex nested values.
+- osx_defaults:
+    domain: com.apple.finder
+    key: ComputerViewSettings:ListViewSettings
+    value:
+        - iconSize: 16
+        - sortColumn: name
+        - textSize: 12
+        - columns:
+            - comments:
+                - ascending: true
 '''
 
+
+from base64 import b64decode
+import calendar
+import contextlib
 import datetime
+import os
 import re
+import string
 
 from ansible.module_utils.basic import AnsibleModule
-from ansible.module_utils.six import binary_type, text_type
-
-
-# exceptions --------------------------------------------------------------- {{{
-class OSXDefaultsException(Exception):
-    pass
-
-
-# /exceptions -------------------------------------------------------------- }}}
-
-# class MacDefaults -------------------------------------------------------- {{{
-class OSXDefaults(object):
-    """ Class to manage Mac OS user defaults """
+from ansible.module_utils.six import string_types, binary_type, text_type
+from ansible.module_utils.six import integer_types
+
+try:
```

Comment:
To ensure pyObjC lib is available even if the user has a homebrewed python version maybe add something like this: ```sys.path.insert(0, '/System/Library/Frameworks/Python.framework/Versions/2.7/Extras/lib/python/PyObjc')```

Reply:
By default, Ansible use the python located at `/usr/bin/python`, so, in macOS, this python is the system provided python that already have PyObjC. No problem in this case.

If someone specify in Ansible another python to use (like a homebrewed python3) I think it's his responsibility to insure that this python have PyObjC installed (just adding the system PyObjC to path will not work with another python version anyway).

Does this make sense?